compiler/control/CompileMethod.cpp

```diff
@@ -276,14 +282,35 @@ compileMethodFromDetails(
    uint64_t translationStartTime = TR::Compiler->vm.getUSecClock();
    OMR::FrontEnd &fe = OMR::FrontEnd::singleton();
    auto jitConfig = fe.jitConfig();
+
+#if defined(OLD_MEMORY)
    TR::RawAllocator rawAllocator;
    TR::SystemSegmentProvider defaultSegmentProvider(1 << 16, rawAllocator);
+
+   // used when debugging memory allocators: remaps allocated segments rather than freeing to trap on use-after-free
    TR::DebugSegmentProvider debugSegmentProvider(1 << 16, rawAllocator);
-   TR::SegmentAllocator &scratchSegmentProvider =
+
+   TR::SegmentAllocator &segmentProvider =
       TR::Options::getCmdLineOptions()->getOption(TR_EnableScratchMemoryDebugging) ?
          static_cast<TR::SegmentAllocator &>(debugSegmentProvider) :
          static_cast<TR::SegmentAllocator &>(defaultSegmentProvider);
-   TR::Region dispatchRegion(scratchSegmentProvider, rawAllocator);
+#else
+   TR::RawAllocator defaultRawAllocator;
+   TR::DebugRawAllocator debugRawAllocator;
+   TR::RawAllocator & rawAllocator =
+      TR::Options::getCmdLineOptions()->getOption(TR_EnableScratchMemoryDebugging) ?
+      debugRawAllocator : defaultRawAllocator;
+
+   TR::BackingMemoryAllocator backingMemoryAllocator(rawAllocator, 1 << 16);;
+
+   TR::DebugSegmentAllocator debugSegmentAllocator(backingMemoryAllocator, 1 << 16);
+   TR::SegmentAllocator defaultSegmentAllocator(backingMemoryAllocator, 1 << 16);
+   TR::SegmentAllocator &segmentAllocator =
+      TR::Options::getCmdLineOptions()->getOption(TR_EnableScratchMemoryDebugging) ?
+      debugSegmentAllocator : defaultSegmentAllocator;
+#endif
+
+   TR::Region dispatchRegion(segmentAllocator, rawAllocator);
```

Comment:
`segmentAllocator` isn't defined when using `OLD_MEMORY`; I guess the variable is named `segmentProvider`.

Reply:
Given the change below (ie using `segmentProvider` for `OLD_MEMORY`, and `segmentAllocator` for the new stuff), it seems like you're going to have to duplicate the initialization of the `TR::Region` in each of the `ifdef` sections